File: third-maximum-number/thirdMaximumNumber.py
"""
Given a non-empty array of integers, return the third maximum number in this array. If it does not exist, return the maximum number. The time complexity must be in O(n).
"""

class Solution(object):
    def thirdMax(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        l = []
        for n in set(nums):
            l = sorted(l + [-n])[:3]
            #bisect.insort(l + [-n])
            #del l[3:]
        return -l[2] if len(l)>2 else -l[0]

File: third-maximum-number/thirdMaximumNumber.java
public class Solution {
    public int thirdMax(int[] nums) {
        TreeSet<Integer> ts = new TreeSet<>();
        for(int i : nums) {
            ts.add(i);
            if(ts.size() > 3) {
                ts.remove(ts.first());
            }
        }
        return ts.size() < 3 ? ts.last() : ts.first();
    }
}